doc/plugin-api-internals.md

```diff
@@ -0,0 +1,488 @@
+     }
+
+     template<typename HookTask__,
+              typename... Args__>
+     static void apply() {
+        HookTask__::invoke(ExamplePlugin);
+        HookTask__::invoke(ExamplePlugin);
+     }
+  };
+}
+
+namespace kaleidoscope {
+  void Hooks::init() {
+    kaleidoscope_internal::OrderedPlugins::template apply<HookTask_init>();
+  }
+
+  bool Hooks::eventHandlerHook(Key &mappedKey,
+                               byte row, byte col, uint8_t keyState) {
+    return kaleidoscope_internal::OrderedPlugins
+              ::template apply<HookTask_eventHandlerHook>
+                            (mappedKey, row, col, keyState);
+  }
+
+  void Hooks::preReportHook() {
+    kaleidoscope_internal::OrderedPlugins::template apply<HookTask_preReportHook>();
+  }
+
+  void Hooks::postReportHook() {
+    kaleidoscope_internal::OrderedPlugins::template apply<HookTask_postReportHook>();
+  }
+}
+```
+
+## Calling
+
+Now that we have the macros expanded, lets see what happens when we call a hook!
+Remember that our entry point is in `Kaleidoscope::Hooks`. We'll first follow
+`Kaleidoscope::Hooks::init()`, then `Kaleidoscope::Hooks::eventHandlerHook`.
+
+### `Kaleidoscope::Hooks::init()`
+
+First, we call `kaleidoscope_internal::OrderedPlugins::template
+apply<HookTask_init>()`. Remember that `apply<HookTask_init>` is a templated
+function, so - at compile time - this expands to:
+
+```c++
+static void apply<HookTask_init>() {
+  HookTask_init::invoke(ExamplePlugin);
+  HookTask_init::invoke(ExamplePlugin);
+}
+```
+
+And `HookTask_init::invoke` expands to:
+
+```c++
+static void invoke(ExamplePlugin_ &plugin) {
+  return plugin.init();
+}
+```
+
+Considering that `HookTask_init` can be inlined, our `apply<HookTask_init>` function is
+effectively this now:
+
+```c++
+static void apply<HookTask_init>() {
+  ExamplePlugin.init();
+  ExamplePlugin.init();
+}
+```
+
+Since this can also be inlined, our `Kaleidoscope::Hooks::init` becomes:
+
+```c++
+void Hooks::init() {
+  ExamplePlugin.init();
+  ExamplePlugin.init();
+}
+```
+
+### `Kaleidoscope::Hooks::eventHandlerHook()`
+
+The only difference here is that we expand `apply<HookTask_eventHandlerHook>` differently:
+
+```c++
+static auto apply(Args__&&... hook_args)
+                        -> typename HookTask__::ReturnType {
+
+   typedef typename HookTask__::ContinuationPredicate
+      ContinuationPredicate;
+
+   typename HookTask__::ReturnType hook_return_val;
+
+   hook_return_val = HookTask__::invoke(ExamplePlugin, hook_args...);
+   if (!ContinuationPredicate::eval(hook_return_val)) {
+     return hook_return_val;
+   }
+
+   hook_return_val = HookTask__::invoke(ExamplePlugin, hook_args...);
+   if (!ContinuationPredicate::eval(hook_return_val)) {
+     return hook_return_val;
+   }
+
+   return hook_return_val;
+}
+```
+
+Which in turn becomes:
+
+```c++
+static auto apply(Key &mappedKey, byte row, byte col, uint8_t keyState) {
+  ContinueIfHookReturnsTrue hook_return_val;
```

Comment:
`ContinueIfHookReturnsTrue` is actually a class, better replace it with `bool` here.

Reply:
Most precise would be 
```cpp
decltype(ContinueIfHookReturnsTrue::eval(true)) hook_return_val;
```
but that's of no use here.